proposals/0530-strict-box-denesting.rst

```diff
@@ -0,0 +1,527 @@
+are maintained.  (The workaround of monomorphising containers even has
+runtime cost, since it necessarily increases code size.)
+
+..
+  There are a few less essential limitations in today's GHC as well.
+  A constructor field of an unlifted type is always treated as a lazy
+  field and thus not unpacked, even with an explicit {-# UNPACK #-}
+  pragma.  Data types that capture an existential type variable are in
+  the same way ineligible for unpacking, and are also ignored by CPR
+  analysis.
+
+..
+  TODO: perhaps add links to past conversations demonstrating demand
+
+
+
+
+Proposed Change Specification
+-----------------------------
+..
+  This section shares a lot with the introduction.
+  When editing one, it is wise to consider editing the other!
+
+This proposal introduces a new optimization flag, tentatively named
+``-fstrict-box-denesting``, which will be enabled by default with
+``-O1``.  When this flag is active while compiling a ``data`` or
+``data instance`` declaration, its constructors are tested for
+"denestability." A constructor is considered denestable when it
+satisfies these four conditions:
+
+..
+  These are identical to the four conditions given in the introduction.
+
+1. There is no other constructor in the relevant ``data`` or
+   ``data instance`` declaration.
+2. Among the constructor's fields (after unpacking, counting
+   existentials and constraints), exactly one field has
+   non-zero-width.
+3. That field's type is represented by a GC-managed pointer.
+4. If that field is of lifted type, it must be a strict field.
+
+A constructor determined to be denestable will not generate any code
+at any of its use sites (even use-sites in modules for which
+``-fstrict-box-denesting`` is disabled), so that using or
+pattern-matching against this constructor has no runtime cost, except
+to the extent that doing so forces evaluation to take place.
+
+Whether a constructor is determined denestable or not has no effect on
+the surface-language semantics of that constructor.  In particular,
+despite the similarity to newtype constructors, a denestable
+constructor does not introduce ``Coercible`` instances and
+pattern-matching a value against a denestable constructor does force
+evaluation of that value.
+
+Additionally, the type of the somewhat obscure ``dataToTag#``
+primitive (and that of its alias ``GHC.Base.getTag``) will be changed
+as follows:
+
+::
+
+   -- Today,
+   dataToTag#, getTag
+     :: a -> Int#
+
+   -- Under this proposal,
+   dataToTag#, getTag
+     :: forall {l :: Levity} (a :: TYPE (BoxedRep l))
+     .  DataToTag a => a -> Int#
+
+``DataToTag`` is a new built-in class with special solving behavior,
+similar to existing special behavior for ``Coercible`` and ``WithDict``.
+Specifically, the constraint ``DataToTag t`` is always soluble when
+``t`` is an algebraic data type (i.e. ``t`` matches a ``data`` or
+``data instance`` declaration) with all of its constructors in scope.
```

Comment:
Yes, dictionary-passing will be necessary in some polymorphic situations. But do any "real" uses of `dataToTag#` exist where the type `a` isn't a statically known ADT? I'm certainly not aware of any. (I expect more breakage due to removing it from `GHC.Prim` as part of the implementation...)

Reply:
It's a bit hard to say, actually. It might occur in insufficiently inlined generic code, for example. But then an additional `DataToTag` constraint is the least of our problems.

Perhaps it's fine. You should specify where you want `DataToTag` to live, though, and that you'll re-export it through `GHC.Exts` (I think?).